ndk/src/font.rs

```diff
@@ -0,0 +1,401 @@
+//! [`AFontMatcher`]: https://developer.android.com/ndk/reference/group/font#afontmatcher_create
+//! [`ASystemFontIterator`]: https://developer.android.com/ndk/reference/group/font#asystemfontiterator_open
+
+use std::convert::TryFrom;
+use std::ffi::{CStr, OsStr};
+use std::fmt;
+use std::os::unix::prelude::OsStrExt;
+use std::path::Path;
+use std::ptr::NonNull;
+
+/// An integer holding a valid font weight value between 1 and 1000.
+///
+/// See the followings for more details:
+/// * [`AFONT_WEIGHT_*`]
+/// * [`Font::weight`]
+///
+/// [`AFONT_WEIGHT_*`]: https://developer.android.com/ndk/reference/group/font#anonymous-enum-33
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct FontWeight(u16);
+
+impl FontWeight {
+    pub const fn new(value: u16) -> std::result::Result<Self, TryFromU16Error> {
+        if 0 < value && value <= 1000 {
+            Ok(Self(value))
+        } else {
+            Err(TryFromU16Error(()))
+        }
+    }
+
+    pub const fn value(&self) -> u16 {
+        self.0
+    }
+
+    pub const THIN: FontWeight = FontWeight(ffi::AFONT_WEIGHT_THIN as u16);
+    pub const EXTRA_LIGHT: FontWeight = FontWeight(ffi::AFONT_WEIGHT_EXTRA_LIGHT as u16);
+    pub const LIGHT: FontWeight = FontWeight(ffi::AFONT_WEIGHT_LIGHT as u16);
+    pub const NORMAL: FontWeight = FontWeight(ffi::AFONT_WEIGHT_NORMAL as u16);
+    pub const MEDIUM: FontWeight = FontWeight(ffi::AFONT_WEIGHT_MEDIUM as u16);
+    pub const SEMI_BOLD: FontWeight = FontWeight(ffi::AFONT_WEIGHT_SEMI_BOLD as u16);
+    pub const BOLD: FontWeight = FontWeight(ffi::AFONT_WEIGHT_BOLD as u16);
+    pub const EXTRA_BOLD: FontWeight = FontWeight(ffi::AFONT_WEIGHT_EXTRA_BOLD as u16);
+    pub const BLACK: FontWeight = FontWeight(ffi::AFONT_WEIGHT_BLACK as u16);
+    pub const MAX: FontWeight = FontWeight(ffi::AFONT_WEIGHT_MAX as u16);
+}
+
+impl fmt::Display for FontWeight {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            FontWeight::THIN => writeln!(f, "Thin"),
+            FontWeight::EXTRA_LIGHT => writeln!(f, "Extra Light (Ultra Light)"),
+            FontWeight::LIGHT => writeln!(f, "Light"),
+            FontWeight::NORMAL => writeln!(f, "Normal (Regular)"),
+            FontWeight::MEDIUM => writeln!(f, "Medium"),
+            FontWeight::SEMI_BOLD => writeln!(f, "Semi Bold (Demi Bold)"),
+            FontWeight::BOLD => writeln!(f, "Bold"),
+            FontWeight::EXTRA_BOLD => writeln!(f, "Extra Bold (Ultra Bold)"),
+            FontWeight::BLACK => writeln!(f, "Black (Heavy)"),
+            _ => writeln!(f, "{}", self.0),
+        }
+    }
+}
+
+/// The error type returned when an invalie font weight value is passed.
+#[derive(Debug)]
+pub struct TryFromU16Error(());
+
+impl fmt::Display for TryFromU16Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            fmt,
+            "font weight must be positive and less than or equal to 1000"
```

Comment:
I tend to prefer `f.write_str()` in these cases.

And:

```suggestion
            "Font weight must be positive and less than or equal to 1000"
```

Reply:
Otherwise, how about printing the faulty value after storing it inside `self`?

Reply:
I'll replace them with `write_str`, but shouldn't error messages start with a lowercase letter? AFAIK, all error messages in `std` start with a lowercase letter.